daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Utf8String.scala

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.digitalasset.daml.lf.data
+
+import java.nio.charset.StandardCharsets
+import java.security.MessageDigest
+
+import scalaz.Equal
+import scalaz.std.string._
+
+import scala.collection.mutable
+
+// The DAML-LF strings are supposed to be UTF-8 while standard java string are UTF16
+// We box standard UTF16 java String to prevent non intentional usages of UTF16 operations
+// (for instance length, charAt, ordering ...) and provide UTF8 emulation methods.
+
+// Use `toString` wisely. As a rule on the thumb, you should use `toString` only for logging,
+// testing, or sending to external libraries (as for instance protobuf/json builders).
+case class Utf8String(override val toString: String) extends AnyVal with Ordered[Utf8String] {
```

Comment:
Two things:

1. `AnyVal` boxes _more_ often, that's why they only really get that benefit for use as extension methods, not so much for newtypes. If you want it to be unboxed, use an abstract member that _doesn't_ have the `<: String` part in the signature.
2. No one can use `toString` wisely.

```suggestion
case class Utf8String(override val javaString: String) extends Ordered[Utf8String] {

  @deprecated("use javaString instead", since = "0.12.18")
  override final def toString = javaString
```

Reply:
OK. Tha's my last `AnyVal` !

Reply:
fixed